src/lib/operators/abstract_read_write_operator.cpp

```diff
@@ -16,7 +16,13 @@ void AbstractReadWriteOperator::execute() {
          "AbstractReadWriteOperator::execute() should never be called without having set the transaction context.");
 
   Assert(_state == ReadWriteOperatorState::Pending, "Operator needs to have state Pending in order to be executed.");
-  AbstractOperator::execute();
+
+  try {
```

Comment:
well, my opinion... I think I can guess your point. Even though there shouldn't be any non-fatal exception arising from an operator, we can't know that for sure and would have to ensure a stable MVCC state here, just in case we'll actually try to recover from the exception that is passing through here??
(Even though we probably never will do that.)
Is it that why you're doing this?

Reply:
This is not about fatal vs non-fatal or about recovering. I wrote the test checking that insert fails when I insert `NULL` into a non-nullable column. Before the change, this would happen:

```
[ RUN      ] OperatorsInsertTest.InsertNullIntoNonNull
libc++abi.dylib: terminating with uncaught exception of type std::logic_error: src/lib/concurrency/transaction_context.cpp:40 Has registered operators but has neither been committed nor rolled back.
Abort trap: 6
```

This is an exception that comes from `~TransactionContext`. So while the test succeeded and the insert operator threw an exception, it left the transaction context in an inconsistent state. An assertion in the context would check for the sanity of the context when it is destroyed at the end of the test.

Even worse - when I take out the `EXPECT_THROW`, the same would happen. The exception shown above would cut in line before the `std::logic_error` saying that you can't insert a `NULL`. This would lead people to debug in the wrong place.

I mentioned your name because I seem to remember that you don't like rethrowing exceptions because it makes debugging harder. In this case, I believe it makes sense.

Reply:
But inserting NULL into a non-nullable column is not something we should handle/try to recover from. So why bother?
Or Insert needs to handle NULL-into-non-null better, i.e. not Assert it but abort the transaction and then we're back on well defined paths. What am I missing?

Ah, it's that google test catches all exceptions, including our intended-to-be-fatal ones and from there on things go south. Okay, we're doing this wrong. We can't make hyrise go into intended-to-trigger-fatal-error paths such as inserting NULL into non-NULL-target and then do precisely that in the tests and expect the executable to recover. Note I am not saying that it needs to be a non-recoverable case I am just saying that by Asserting it we're declaring it as one. All our EXPECT_THROW stuff is essentially wrong and should be revisited.

Reply:
Not part of this PR though, I'll make up my mind on this and open an issue.

Reply:
Still, but some comment on the catch-everything clause atrocity pls.

Reply:
done.